test/framework/src/main/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java

```diff
@@ -596,6 +596,81 @@ public static ClusterState stateWithAssignedPrimariesAndReplicas(
         return state.build();
     }
 
+    public static ClusterState stateWithUnassignedPrimariesAndReplicas(String[] indices, int numberOfShards, int numberOfReplicas) {
+        return stateWithUnassignedPrimariesAndReplicas(
+            Metadata.DEFAULT_PROJECT_ID,
+            indices,
+            numberOfShards,
+            Collections.nCopies(numberOfReplicas, ShardRouting.Role.DEFAULT)
+        );
+    }
+
+    /**
+     * Creates cluster state with several indexes, shards and replicas (with given roles) and all shards STARTED.
+     */
+    public static ClusterState stateWithUnassignedPrimariesAndReplicas(
+        ProjectId projectId,
+        String[] indices,
+        int numberOfShards,
+        List<ShardRouting.Role> replicaRoles
+    ) {
+        int numberOfDataNodes = replicaRoles.size() + 1;
+        DiscoveryNodes.Builder discoBuilder = DiscoveryNodes.builder();
+        for (int i = 0; i < numberOfDataNodes + 1; i++) {
+            final DiscoveryNode node = newNode(i);
+            discoBuilder = discoBuilder.add(node);
+            if (i == 0) {
+                discoBuilder.localNodeId(node.getId());
+            } else if (i == numberOfDataNodes) {
+                discoBuilder.masterNodeId(node.getId());
+            }
+        }
```

Comment:
If I'm reading it right, the actual number of data nodes ends up being `numberOfDataNodes + 1`, this seems counter-intuitive, could we change `numberOfDataNodes` to have the actual number of data nodes?

Reply:
Yeah, this is pretty convoluted. I never did find a reason for it in the other helpers. Fixing it 👍